SPEC.md

```diff
@@ -0,0 +1,180 @@
+  benefits of peer-to-peer networking without implementing a full libp2p stack.
+
+### Technical Details
+
+The libp2p daemon and client will communicate with each other over a simple TCP
+protocol built with [protobuf](https://developers.google.com/protocol-buffers/).
+At present, the protocol operates over Unix sockets.
+
+Future implementations may attempt to take advantage of shared memory (shmem)
+or other IPC constructs.
+
+## Protocol Specification
+
+### Data Types
+
+The data structures are defined in [pb/p2pd.proto](pb/p2pd.proto).
+
+### Protocol Requests
+
+*Protocols described in pseudo-go. Items of the form [item, ...] are lists of
+many items.*
+
+#### Errors
+
+Any response that may be an error, will take the form of:
+
+```
+Response{
+  Type: ERROR,
+  ErrorResponse: {
+    Msg: <error message>,
+  },
+}
+```
+
+#### `Connect`
+
+Clients issue a `Connect` request when they wish to add a known peer's addresses
+to the daemon's `Peerstore` and connect to it.
+
+**Client**
+```
+Request{
+  Type: CONNECT,
+  ConnectRequest: {
+    Peer: <peer id>,
+    Addrs: [<addr>, ...],
+  },
+}
+```
+
+**Daemon**
+*May return an error.*
+```
+Response{
+  Type: OK,
+}
+```
+
+#### `Identify`
+
+Clients issue an `Identify` request when they wish to determine the peer ID and
+listen addresses of the daemon.
+
+**Client**
+```
+Request{
+  Type: IDENTIFY,
+}
+```
+
+**Daemon**
+```
+Response{
+  Type: OK,
+  IdentifyResponse: {
+      Id: <daemon peer id>,
+      Addrs: [<daemon listen addr>, ...],
+  },
+}
+```
+
+#### `StreamOpen`
+
+Clients issue a `StreamOpen` request when they wish to initiate an outbound
+stream to a peer on one of a set of protocols.
+
+**Client**
+```
+Request{
+  Type: STREAM_OPEN,
+  StreamOpenRequest: {
+    Peer: <peer id>,
+    Proto: [<protocol string>, ...],
+  },
+}
+```
+
+**Daemon**
+*May return an error, short circuiting.*
+```
+Response{
+  Type: OK,
+  StreamInfo: {
+    Peer: <peer id>,
+    Addr: <peer address connected to>,
+    Proto: <protocol we connected on>,
+  },
+}
+```
+
+After writing the response message to the socket, the daemon begins piping the
+newly created stream to the client over the socket. Clients may read from and
+write to the socket as if it were the stream. **WARNING**: Clients must be
+careful not to read excess bytes from the socket when parsing the daemon
+response, otherwise they risk reading into the stream output.
+
+#### `StreamHandler` - Register
+
+Clients issue a `StreamHandler` request to register a handler for inbound
+streams on a given protocol. Prior to issuing the request, the client must be
+listening to a Unix socket at the specified path.
+
+**Client**
+```
+Request{
+  Type: STREAM_HANDLER,
+  StreamHandlerRequest: {
+    Path: <path to unix socket client is listening to>,
+    Proto: [<protocols to route to this handler>, ...],
+  }
+}
+```
+
+**Daemon**
+*May return an error indicating a handler for a specified protocol was already
+registered.*
```

Comment:
Actually, it just goes forward and overwrites the stream binding. Maybe we should return an error though.